### ISSUE TITLE
Split `clippy::manual_is_power_of_two` into two lints to handle `x & (x - 1) == 0` better

### ISSUE THREAD

**Issue by daira** [rust-lang/rust-clippy#14309]
### What it does

The existing `clippy::manual_is_power_of_two` lint, which triggers on expressions of the form either `x.count_ones() == 1` or `x & (x - 1) == 0` where `x` is of unsigned integer type, was moved from default to `pedantic` in #13553. The rationale stated in #13547 was that `x.count_ones() == 1` might be a clearer way of expressing the intent if `x` represents a bit flag, for example.

The suggestion is to:
* Change `clippy::manual_is_power_of_two` to trigger only on `x.count_ones() == 1`.
* Add, say `clippy::bithack_is_power_of_two` that triggers only on `x & (x - 1) == 0` and similar. The message should make clear that another reason not to use this form is that it is incorrect for `x == 0` (or for `x <= 0` if `x` is of a signed type).

The reason to split into two lints, rather than just have `clippy::manual_is_power_of_two` produce different lint messages, is that it is quite plausible that users would want to configure them differently. For example, the reporter of #13547 would only have needed to suppress `clippy::manual_is_power_of_two`.

Note that in Rust, if `x` is the minimum value of its type then `x - 1` is unspecified behaviour; it may panic in debug mode. So the correct expression would have to be `x > 0 && x & (x - 1) == 0` or equivalent. If the code actually uses this expression then the lint message might need to be different.

If `x` is of a signed type, then `x.is_power_of_2()` doesn't exist. A programmer might still try to use `x > 0 && x & (x - 1) == 0` there &mdash; which is correct, if the intent is only to check for positive powers of 2. For bonus points suggest something like `uN::try_from(x).is_some_and(|u| u.is_power_of_two())` (where `uN` is the appropriate unsigned integer type) in that case.

### Advantage

It's clearer and handles `x <= 0` correctly.

### Drawbacks

There might be reason to use `x & (x - 1) == 0` in code that needs to operate on secret data in constant time and where `x` must be positive. (Either `x.count_ones()` or `x.is_power_of_two()` might be non-constant-time, especially on architectures that have no "popcount" instruction.) That probably doesn't happen very often and in that case I think it would be reasonable to suppress the lint.

If a user had explicitly disabled `clippy::manual_is_power_of_two` during the period when it was default-enabled, and now need to disable `clippy::bithack_is_power_of_two`, they might be irritated. This seems like a relatively minor issue.

### Example

```rust
let x: u32 = ...;
x & (x - 1) == 0
```
warning: reimplementing `is_power_of_two` using a bit hack
> ```rust
> x & (x - 1) == 0
> ^^^^^^^^^^^^^^^^ help: consider using `.is_power_of_two()`: `x.is_power_of_two()`
> ```
= help: Note that `x - 1` is unspecified when `x == 0`. It will panic in debug mode, or give the wrong answer if it was intended to check for a power of two.
= help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bithack_is_power_of_two

----

```rust
let x: u32 = ...;
x > 0 && x & (x - 1) == 0
```
warning: reimplementing `is_power_of_two` using a bit hack
> ```rust
> x > 0 && x & (x - 1) == 0
> ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using `.is_power_of_two()`: `x.is_power_of_two()`
> ```
= help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bithack_is_power_of_two

----

```rust
let x: i32 = ...;
x & (x - 1) == 0
```
warning: reimplementing `is_power_of_two` using a bit hack
> ```rust
> x & (x - 1) == 0
> ^^^^^^^^^^^^^^^^ help: consider converting to u32 and using `.is_power_of_two()`:
> `u32::try_from(x).is_some_and(|u| u.is_power_of_two())``
> ```
= help: Note that `x - 1` is unspecified when `x == i32::MIN`. It will panic in debug mode. Also, this code will give the wrong answer for `x <= 0` if it was intended to check for a power of two.
= help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bithack_is_power_of_two

----

```rust
let x: i32 = ...;
x > 0 && x & (x - 1) == 0
```
warning: reimplementing `is_power_of_two` using a bit hack
> ```rust
> x > 0 && x & (x - 1) == 0
> ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider consider converting to u32 and using `.is_power_of_two()`:
> `u32::try_from(x).is_some_and(|u| u.is_power_of_two())``
> ```
= help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bithack_is_power_of_two
